server/services/team/service.ts

```diff
@@ -1,40 +1,106 @@
-import { Authorization } from '../../sdk';
-import { ITeam } from '../../../definition/ITeam';
+import { UsersRaw } from '../../../app/models/server/raw/Users';
+import { TeamMemberRaw } from '../../../app/models/server/raw/TeamMember';
+import { IRoom } from '../../../definition/IRoom';
 
 export class TeamService extends ServiceClass implements ITeamService {
 	protected name = 'team';
 
-	// constructor(db: Db) {
-	// 	super();
-	// }
+	private TeamModel: TeamRaw;
+
+	private Users: UsersRaw;
+
+	private TeamMembersModel: TeamMemberRaw;
+
+	constructor(db: Db) {
+		super();
 
-	async create(uid: string, doc: Omit<ITeam, '_id'>): Promise<ITeam> {
+		this.TeamModel = new TeamRaw(db.collection('rocketchat_team'));
+		this.TeamMembersModel = new TeamMemberRaw(db.collection('rocketchat_team_member'));
+		this.Users = new UsersRaw(db.collection('users'));
+	}
+
+	async create(uid: string, { team, room, members }: ITeamCreateParams): Promise<ITeam> {
 		const hasPermission = await Authorization.hasPermission(uid, 'create-team');
 		if (!hasPermission) {
 			throw new Error('no-permission');
 		}
 
-		const team = {
-			_id: 'somethi',
-			...doc,
+		const createdBy = await this.Users.findOneById(uid, { projection: { username: 1 } });
+		if (!createdBy) {
+			throw new Error('invalid-user');
+		}
+
+		// TODO add validations to `data` and `members`
+
+		const membersResult = await this.Users.findActiveByIds(members, { projection: { username: 1, _id: 0 } }).toArray();
+		const memberUsernames = membersResult.map(({ username }) => username);
+
+		const teamData = {
+			...team,
+			createdAt: new Date(),
+			createdBy,
+			_updatedAt: new Date(), // TODO how to avoid having to do this?
 		};
 
-		return team;
-	}
+		// create team
+		const result = await this.TeamModel.insertOne(teamData);
+		const teamId = result.insertedId;
 
-	async list(uid: string): Promise<Array<ITeam>> {
-		return [{
-			_id: 'aa',
-			name: 'ok',
-			type: 0,
+		const membersList: Array<Omit<ITeamMember, '_id'>> = members?.map((memberId) => ({
+			teamId,
+			userId: memberId,
+			createdAt: new Date(),
+			createdBy,
+			_updatedAt: new Date(), // TODO how to avoid having to do this?
+		})) || [];
+
+		membersList.push({
+			teamId,
+			userId: uid,
+			roles: ['owner'],
 			createdAt: new Date(),
-			createdBy: {
-				_id: 'rocket.cat',
-				username: 'rocket.cat',
+			createdBy,
+			_updatedAt: new Date(), // TODO how to avoid having to do this?
+		});
+
+		await this.TeamMembersModel.insertMany(membersList);
+
+		const roomType: IRoom['t'] = team.type === TEAM_TYPE.PRIVATE ? 'p' : 'c';
+
+		const newRoom = {
+			...room,
+			type: roomType,
+			name: team.name,
+			members: memberUsernames,
+			extraData: {
+				...room.extraData,
+				teamId,
 			},
-			_updatedAt: new Date(),
-		}];
+		};
+
+		await Room.create(uid, newRoom);
+
+		return {
+			_id: teamId,
+			...teamData,
+		};
+	}
+
+	async list(uid: string): Promise<Array<ITeam>> {
```

Comment:
your approach brought some questions 🤔 

I think we need a way to return *all teams*, so let's say an admin can manage all teams from a server, and I thought this method should return that. But I think we'll need a special permission for that since a regular user should not be able to see private teams he is not part of. wdyt?

Reply:
agree. i've added this logic along with the required permission on my last commit. let me know what you think of it